sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQosImpl.java

```diff
@@ -0,0 +1,909 @@
+     */
+    private double throttlingProbability(Instant instantSinceEpoch) {
+      if (!allRequestsMovingFunction.isSignificant()) {
+        return 0;
+      }
+      long nowMsSinceEpoch = instantSinceEpoch.getMillis();
+      long allRequestsCount = allRequestsMovingFunction.get(nowMsSinceEpoch);
+      long successfulRequestsCount = successfulRequestsMovingFunction.get(nowMsSinceEpoch);
+
+      double overloadMaxCount = overloadRatio * successfulRequestsCount;
+      double overloadUsage = allRequestsCount - overloadMaxCount;
+
+      double calcProbability = overloadUsage / (allRequestsCount + MIN_REQUESTS);
+      allRequestsCountDist.update(allRequestsCount);
+      successfulRequestsCountDist.update(successfulRequestsCount);
+      overloadMaxCountDist.update((long) overloadMaxCount);
+      overloadUsageDist.update((long) (overloadUsage * 100));
+      throttleProbabilityDist.update((long) (calcProbability * 100));
+      return Math.max(0, calcProbability);
+    }
+  }
+
+  /**
+   * An implementation providing the 500/50/5 ramp up strategy recommended by <a target="_blank"
+   * rel="noopener noreferrer"
+   * href="https://cloud.google.com/firestore/docs/best-practices#ramping_up_traffic">Ramping up
+   * traffic</a>.
+   */
+  @VisibleForTesting
+  static final class WriteRampUp {
+
+    private static final Duration RAMP_UP_INTERVAL = Duration.standardMinutes(5);
+    private final int baseBatchBudget;
+    private final long rampUpIntervalMinutes;
+    private final MovingFunction writeCounts;
+    private final LinearBackoff backoff;
+    private final Distribution throttlingMs;
+    private final Distribution availableWriteCountBudget;
+
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    private Optional<Instant> firstInstant = Optional.empty();
+
+    WriteRampUp(int baseBatchBudget, DistributionFactory distributionFactory) {
+      this.baseBatchBudget = baseBatchBudget;
+      this.rampUpIntervalMinutes = RAMP_UP_INTERVAL.getStandardMinutes();
+      this.writeCounts =
+          createMovingFunction(
+              // track up to one second of budget usage.
+              //   this determines the full duration of time we want to keep track of request counts
+              Duration.standardSeconds(1),
+              // refill the budget each second
+              //   this determines the sub-granularity the full duration will be broken into. So if
+              //   we wanted budget to refill twice per second, this could be passed
+              //   Duration.millis(500)
+              Duration.standardSeconds(1));
+      this.backoff = new LinearBackoff(Duration.standardSeconds(1));
+      this.throttlingMs =
+          distributionFactory.get(RpcQos.class.getName(), "qos_rampUp_throttlingMs");
+      this.availableWriteCountBudget =
+          distributionFactory.get(RpcQos.class.getName(), "qos_rampUp_availableWriteCountBudget");
+    }
+
+    int getAvailableWriteCountBudget(Instant instant) {
```

Comment:
Correct, this only tracks data for up to one second. In the case of this ramp up calculation we're not taking past performance into consideration (instead the write batcher and adaptive throttler take care of that). This class is only attempt to limit based on the 555 budget and what has been used within that specific one-second window.

The budget value grows relative to the `firstInstant` captured, so even if a worker is only sending a request every 5 seconds the budget calculation will still be for the next time window. If the worker is sending more than one request a second, then it will decrement from the budget and be throttled until the next second if budget has been exhausted for the specific second.

For `hintMaxNumWorkers = 1` the budget will fill along this [line](https://www.wolframalpha.com/input/?i=500+*+1.5%5Emax%280%2C+%28x-5%29%2F5%29+from+0+to+30).

I've got a test `org.apache.beam.sdk.io.gcp.firestore.RpcQosSimulationTest#writeRampUp_shouldScaleAlongTheExpectedLine` which runs a series of operations against the RampUp to ensure values fall on the expected line.

Reply:
This is a screenshot of a job that ran for ~90min where we can see the growth steps for the the ramp up budget (the teal line/square line). Since this is aggregated across all workers it's not as exact as each of the 5 minute windows a single worker would see, but does show the progression.

![qos batch size calculation](https://user-images.githubusercontent.com/1754037/117369481-f7283e00-ae92-11eb-9ed1-b26caf51b27d.png)

For the same job we can see the actual (aggregated) min/mean/max batchCapacityCount over the duration of the job. We can see the steps pretty clearly on the left before capping out at 500, while at the same time the ramp up budget keeps increasing all the way up to almost 20k by the end of the job.

![qos_write_batchCapacityCount distribution](https://user-images.githubusercontent.com/1754037/117369946-a6fdab80-ae93-11eb-8c1c-15376551a11d.png)